Create Add to Cart With Options wrapper block

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR creates an `add-to-cart-with-options` wrapper block.

![Screenshot 2024-11-26 at 7 33 45 PM](https://github.com/user-attachments/assets/372854fb-59b1-4bf3-8922-d3635f83a24a)

Closes #53008.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate the `WooCommerce Beta Tester` plugin
2. Go to `WooCommerce Admin Test Helper` -> `Features` and enable `blockified-add-to-cart`.
3. Create a new product
4. Go to Pages > Add New Page
5. Add a `Single product` block and select the product you just created.
6. Go to the block inserter and add an `Add to Cart with Options (Experimental)` block.
7. It should look and behave exactly like the `Add to Cart with Options` block.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Create AddToCartWithOptions

octaedro committed:
Add changelog

octaedro committed:
remove comment

octaedro committed:
Remove changes in single-product

retrofox reviewed (commented):
Hi, Fernando. Thanks for taking over this tough one.
This PR adds many changes we will need to address in the future but requires more relevant, even mandatory, requirements.

For instance, the content of the blocks must be based on children's blocks. This is the primary purpose of this **wrapper** block:

<img width="350" alt="image" src="https://github.com/user-attachments/assets/307c369a-56d1-4972-8412-d124b8291f7e">

Also, if I'm not wrong, we said moving the folder location out of the `product-elements` folder (why this folder?).

Or, I don't see the script running on the client/frontend side.

<img width="376" alt="image" src="https://github.com/user-attachments/assets/c9779605-3aa1-43e8-bfdb-5da3568aece2">

I suggest making this PR smaller, tackling the most important requirements.
What do you think?

octaedro committed:
Move folder and trim component

octaedro commented:
Hi @retrofox,
As we discussed in Slack, I changed the approach for this PR and trimmed a lot of things in the commit f301fe6.
I also moved the folder location.

> Or, I don't see the script running on the client/frontend side.

This is interesting 🤔. I'm seeing the files correctly on my end and in JN.

![Screenshot 2024-11-27 at 10 53 31 AM](https://github.com/user-attachments/assets/2b229d04-e61f-448d-a5ed-2000b7e30518)

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/block.json:
```diff
@@ -0,0 +1,26 @@
+			"default": "input"
+		}
+	},
+	"keywords": [ "WooCommerce" ],
```
I'm not sure we need this keyword

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,92 @@
+const AddToCartFormEdit = ( props: BlockEditProps< Attributes > ) => {
+	const { setAttributes } = props;
+
+	const isStepperLayoutFeatureEnabled = getSettingWithCoercion(
```
The stepper feature doesn't belong to this block. It belongs to the quantity selector block [#53009](https://github.com/woocommerce/woocommerce/issues/53009)

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/settings.tsx:
```diff
@@ -0,0 +1,116 @@
+/**
```
It should belong to the quantity block [#53009](https://github.com/woocommerce/woocommerce/issues/53009)

retrofox reviewed (commented):
This PR still contain logic that shoulnd't be here, especially what belongs to the Quantity block. I think we need to keep 🧹

retrofox commented:
> Hi @retrofox, As we discussed in Slack, I changed the approach for this PR and trimmed a lot of things in the commit [f301fe6](https://github.com/woocommerce/woocommerce/commit/f301fe61b0bc7e66de5c448231cf9cef1fdd9936). I also moved the folder location.
> 
> > Or, I don't see the script running on the client/frontend side.

Nice, thanks for taking over of it.
 
> This is interesting 🤔. Since we weren't using that anymore, I removed the `frontend.ts` but I can bring it back to life :-)

Definitely we'll need this file, but we can add it in a follow-up.

octaedro committed:
Clean block.json

octaedro committed:
clean edit.tsx

octaedro committed:
clean settings.tsx

octaedro committed:
Clean AddToCartWithOptions

octaedro committed:
remove quantitySelectorStyle

octaedro committed:
Add frontend.ts

octaedro committed:
remove isStepperLayoutFeatureEnabled

octaedro committed:
Add comment

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/style.scss:
```diff
@@ -0,0 +1 @@
+/* Styles here */
```
Added this comment here just to avoid the `no-empty-source` lint error

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -0,0 +1,9 @@
+/**
+ * External dependencies
+ */
+import { store } from '@woocommerce/interactivity';
+
+store( 'woocommerce/add-to-cart-with-options', {
+	state: {},
+	actions: {},
+} );
```
I added this file here just because we will need it later.

retrofox commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -0,0 +1,210 @@
+	 */
+	protected function enqueue_data( array $attributes = [] ) {
+		parent::enqueue_data( $attributes );
+		$this->asset_data_registry->add( 'isBlockifiedAddToCart', Features::is_enabled( 'blockified-add-to-cart' ) );
```
no relevant to this PR, but I think the features should be exposed by another process.

octaedro replied:
+1 to this! Should we create an issue for that? Are you thinking about something similar to what we do in [the frontend](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/Features.php#L320)?

retrofox replied:
That's a nice idea. or can we perform an async request?
Feel free to create an issue, but this is out of the scope of this block project

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/index.tsx:
```diff
@@ -0,0 +1,35 @@
+ * Internal dependencies
+ */
+import metadata from './block.json';
+import AddToCartFormEdit from './edit';
```
should we use the `AddToCartOptionsEdit` name?

```suggestion
import AddToCartOptionsEdit from './edit';
```

octaedro replied:
@retrofox Done! I also renamed `AddToCartFormSettings` to `AddToCartOptionsSettings`.

octaedro committed:
Update plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/index.tsx

Co-authored-by: Damián Suárez <rdsuarez@gmail.com>

octaedro committed:
Rename a couple of consts

retrofox commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -0,0 +1,210 @@
+	protected function initialize() {
+		parent::initialize();
+		add_filter( 'wc_add_to_cart_message_html', array( $this, 'add_to_cart_message_html_filter' ), 10, 2 );
+		add_filter( 'woocommerce_add_to_cart_redirect', array( $this, 'add_to_cart_redirect_filter' ), 10, 1 );
```
Should it be addressed here or at the Button level? @Aljullu?

Aljullu replied:
Good question. Probably at the Button level it would be better, no? But I'm fine leaving it here for now and moving it to the Button block in the follow-up PR.

octaedro replied:
Let's tackle this as part of issue https://github.com/woocommerce/woocommerce/issues/53010

retrofox commented on plugins/woocommerce-blocks/bin/webpack-entries.js:
```diff
@@ -19,6 +19,9 @@ const blocks = {
 		customDir: 'product-elements/add-to-cart-form',
 	},
+	'add-to-cart-with-options': {
+		customDir: 'add-to-cart-with-options',
```
Set as experimental here. This is a PR blocker 🚫

octaedro replied:
Added in the commit [53422cf](https://github.com/woocommerce/woocommerce/pull/53215/commits/53422cfcb5588296187504580915e644b60e3c94)

octaedro committed:
Add isExperimental feature flag

retrofox commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -0,0 +1,210 @@
+			)
+		);
+
+		$form = sprintf(
+			'<div %1$s %2$s>%3$s</div>',
+			$wrapper_attributes,
+			'',
+			$product_html
+		);
```
We should try to avoid markup manipulation as much as possible. Let's use the HTML API.
Not a blocker, but something to change in the short time, imo

Aljullu replied:
For reference, issue created here: https://github.com/woocommerce/woocommerce/issues/53282.

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/block.json:
```diff
@@ -0,0 +1,20 @@
+{
+	"name": "woocommerce/add-to-cart-with-options",
+	"version": "1.0.0",
+	"title": "Add to Cart with Options (Blockified)",
```
I'd continue using `Experimental` here.

octaedro replied:
Done in [261e9e6](https://github.com/woocommerce/woocommerce/pull/53215/commits/261e9e6ad101eb7688bec495ea318bac99a5b500)

Aljullu commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,67 @@
+export type UpdateFeaturesType = ( key: FeaturesKeys, value: boolean ) => void;
+
+// Pick the value of the "blockify add to cart flag"
+const isBlockifiedAddToCart = getSettingWithCoercion(
```
Do we need this? My assumption is that if this flag is false, this block shouldn't be registered at all. So at the point we arrive at this code path, `isBlockifiedAddToCart` would always be true, no?

octaedro replied:
Done in [4f9edfb](https://github.com/woocommerce/woocommerce/pull/53215/commits/4f9edfb92f3968a7c0c14b01e2313638a1c992b1)

Aljullu commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,67 @@
+	const { setAttributes } = props;
+
+	const blockProps = useBlockProps( {
+		className: 'wc-block-add-to-cart-with-options-wrapper',
```
What do you think about removing the `-wrapper` part? If the block is called _Add to Cart with Options_, I would keep the CSS class in sync.

Aljullu commented on plugins/woocommerce-blocks/bin/webpack-entries.js:
```diff
@@ -19,6 +19,10 @@ const blocks = {
 		customDir: 'product-elements/add-to-cart-form',
 	},
+	'add-to-cart-with-options': {
+		customDir: 'add-to-cart-with-options',
```
If the directory path matches the block slug, I think `customDir` is not needed.

octaedro replied:
Done in [fa2ac3c](https://github.com/woocommerce/woocommerce/pull/53215/commits/fa2ac3c43e5bb12e2ca036651f3b6b365e7cbcd3)

Aljullu commented on plugins/woocommerce/changelog/add-53008_add_to_cart_wrapper_block:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Create Add to Cart With Options wrapper block #53215
```
Should this be a comment instead of a changelog entry, considering it's an experimental block?

octaedro replied:
Done in [480903e](https://github.com/woocommerce/woocommerce/pull/53215/commits/480903e6247b3da21814503b03082d2ba726fc8f)

Aljullu commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,67 @@
+			blockClientId: blockProps?.id,
+		} );
+
+	useEffect( () => {
+		setAttributes( {
+			isDescendentOfSingleProductBlock,
+		} );
+	}, [ setAttributes, isDescendentOfSingleProductBlock ] );
```
I know this comes from the previous block, but I think setting attributes after the first render should be considered an anti-pattern. It causes some weird bugs in the Site Editor (https://github.com/woocommerce/woocommerce/issues/48936) and it also makes the usage of patterns problematic, as a pattern could be used in both contexts.

So I don't think we should block this PR because of this, but I wonder if we should have a follow-up issue to make sure we refactor this before releasing. What do you think?

retrofox replied:
Maybe we can try to find a different way to handle this instead of using a block attribute.
The main reason is that we need to detect it on the server side, right?

Aljullu replied:
Yeah, that's my understanding as well. I don't know yet which alternative approach we could try, but I would love to find another way. :smile:

octaedro replied:
I just created the issue [53283](https://github.com/woocommerce/woocommerce/issues/53283) to tackle this problem.

Aljullu reviewed (commented):
Great work, @octaedro! I left some minor inline comments below.

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/block.json:
```diff
@@ -0,0 +1,20 @@
+	"description": "Display a button so the customer can add a product to their cart. Options will also be displayed depending on product type. e.g. quantity, variation.",
+	"category": "woocommerce-product-elements",
+	"attributes": {
+		"isDescendentOfSingleProductBlock": {
```
I think we should try to find a better way to detect when the block is a descendent of a Single Product block.
It pollutes the block.

octaedro replied:
I created the issue https://github.com/woocommerce/woocommerce/issues/53284 to tackle this problem.

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,67 @@
+	const blockProps = useBlockProps( {
+		className: 'wc-block-add-to-cart-with-options-wrapper',
+	} );
+	const { isDescendentOfSingleProductBlock } =
+		useIsDescendentOfSingleProductBlock( {
+			blockClientId: blockProps?.id,
+		} );
+
+	useEffect( () => {
+		setAttributes( {
+			isDescendentOfSingleProductBlock,
+		} );
+	}, [ setAttributes, isDescendentOfSingleProductBlock ] );
```
related to the block attributes comment above, I'd like to avoid this way to identify when the block instance is a descendant of the SingleProduct block.

octaedro replied:
I created the issue https://github.com/woocommerce/woocommerce/issues/53284 to tackle this problem.

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypesController.php:
```diff
@@ -368,6 +368,7 @@ protected function get_block_types() {
 		$block_types = array(
 			'ActiveFilters',
 			'AddToCartForm',
+			'AddToCartWithOptions',
```
Should this be moved under `if ( Features::is_enabled( 'experimental-blocks' ) ) { ... }` and specifically under an `if` that checks the feature flag `isBlockifiedAddToCart`?

octaedro replied:
Done in [b4fea60](https://github.com/woocommerce/woocommerce/pull/53215/commits/b4fea60e432c0779583ab09ac345ee830cb3da04)

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx:
```diff
@@ -0,0 +1,67 @@
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+import { BlockEditProps } from '@wordpress/blocks';
+import { getSettingWithCoercion } from '@woocommerce/settings';
+import { isBoolean } from '@woocommerce/types';
```
can you set the var as type explicitly?

```suggestion
import type { isBoolean } from '@woocommerce/types';
```

octaedro committed:
Move feature flag

octaedro committed:
remove customDir from webpack

octaedro committed:
Rename block

octaedro committed:
clean edit.tsx

octaedro committed:
Remove not used imports

octaedro committed:
Fix changelog

octaedro committed:
Fix changelog

octaedro commented:
Thank you, @retrofox and @Aljullu, for your suggestions.
I just addressed them. Could you take another look at this PR?

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypesController.php:
```diff
@@ -464,6 +464,9 @@ protected function get_block_types() {
 			$block_types[] = 'ProductFilterClearButton';
 			$block_types[] = 'ProductFilterCheckboxList';
 			$block_types[] = 'ProductFilterChips';
+			if ( Features::is_enabled( 'isBlockifiedAddToCart' ) ) {
```
My bad, the feature flag key is actually `blockified-add-to-cart`:
```suggestion
			if ( Features::is_enabled( 'blockified-add-to-cart' ) ) {
```

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -0,0 +1,210 @@
+use Automattic\WooCommerce\Blocks\Utils\StyleAttributesUtils;
+
+/**
+ * CatalogSorting class.
```
This needs to be updated. :slightly_smiling_face:

octaedro replied:
Good eyes! Fixed on [f60e11d](https://github.com/woocommerce/woocommerce/pull/53215/commits/f60e11d468e449ee1a2f6f76cbd009a478a827f0)

Aljullu reviewed (approved):
Thanks for the updates, @octaedro! Besides the couple of comments I left, this is looking great to me, so approving from my side.

octaedro committed:
Fix comment

octaedro committed:
Fix feature name

helgatheviking commented:
@octaedro Excited to see this begin! I just pulled this down locally and I don't think the frontend.ts script is getting loaded? I think this is because core [iAPI stores](https://github.com/woocommerce/woocommerce/pull/53215/files#diff-a0064c1e28a09f52947e6d5be8c43730b54d077082989448215c3e455def0c3e) needs to be loaded via `wp_enqueue_script_module`

There's probably a need to handle custom webpack as well. It seems like there is some functionality in `AbstractBlock` that could probably be handled by WP's handling of `block.json`
